Mathlib/Algebra/Star/Positive.lean

```diff
@@ -0,0 +1,178 @@
+nonnegative elements, the "positive" terminology is standard throughout the literature (at least
+in C⋆-algebra theory and operator theory), so we choose to prefer that.
+
+The advantage of using this over other definitions of positivity is that it allows us to unify
+several different concepts under a single umbrella. For example, one might be tempted to consider
+the collection of bounded linear operators on an Hilbert space and define positivity as
+`∀ x, 0 ≤ ⟪T x, x⟫`, but this doesn't generalize nicely to C⋆-algebras. Alternatively, one could
+define positivity as those self-adjoint elements with nonnegative spectrum, but this requires the
+algebra to be unital. Our definition herein conveniently sidesteps these issues.
+
+We place everything in the `StarOrderedRing` namespace in order to avoid collisions or confusion
+with other parts of the library.
+
+## Main definitions
+
+* `StarOrderedRing.positive R`: The additive submonoid of a star ordered ring `R` generated by
+  elements of the form `star s * s` for `s : R`.
+* `StarOrderedRing.positive.selfAdjoint : positive R →+ selfAdjoint R`: the identity map as an
+  `AddMonoidHom`.
+* `StarOrderedRing.positive.instAddMonoidWithOne`: Makes `positive R` into an `AddMonoidWithOne`
+* `StarOrderedRing.positive.instCanonicallyOrderedAddMonoid`: Makes `positive R` into a
+  `CanonicallyOrderedAddMonoid` which avoids duplication of some lemmas (e.g., `zero_le`)
+
+-/
+
+universe u
+
+variable {R : Type u}
+
+namespace StarOrderedRing
+
+def positive (R : Type u) [NonUnitalSemiring R] [PartialOrder R] [StarOrderedRing R] :
+    AddSubmonoid R :=
+  AddSubmonoid.closure (Set.range fun s : R => star s * s)
+
+section NonUnitalSemiring
+
+variable [NonUnitalSemiring R] [PartialOrder R] [StarOrderedRing R]
+
+theorem positive_def {x : R} :
+    x ∈ positive R ↔ x ∈ AddSubmonoid.closure (Set.range fun s : R => star s * s) :=
+  Iff.rfl
+
+theorem positive_iff {x : R} : x ∈ positive R ↔ 0 ≤ x := by rw [positive_def, nonneg_iff]
+
+theorem positive_coe_set : (positive R : Set R) = {x | 0 ≤ x} :=
+  Set.ext fun _x => positive_iff
+
+theorem le_iff_exists_positive {x y : R} : x ≤ y ↔ ∃ p : positive R, x + p = y := by
+  rw [le_iff]
+  constructor
+  · rintro ⟨p, hp, rfl⟩
+    exact ⟨⟨p, hp⟩, rfl⟩
+  · rintro ⟨p, rfl⟩
+    exact ⟨p, p.prop, rfl⟩
+
+theorem positive.coe_le_coe {x y : positive R} : (x : R) ≤ y ↔ x ≤ y :=
+  Iff.rfl
+
+instance positive.instCanonicallyOrderedAddMonoid : CanonicallyOrderedAddMonoid (positive R) :=
+  { inferInstanceAs (OrderedAddCommMonoid (positive R)) with
+    bot := 0
+    bot_le := fun x => positive_iff.mp x.prop
+    exists_add_of_le := @fun x y (h : (x : R) ≤ y) => by
+      convert le_iff_exists_positive.mp h using 1
+      funext p
+      rw [@eq_comm _ y]
+      exact_mod_cast rfl
+    le_self_add := fun x y => show (x : R) ≤ x + y from le_iff_exists_positive.mpr ⟨y, rfl⟩ }
+
+theorem positive.isSelfAdjoint (x : positive R) : IsSelfAdjoint (x : R) := by
```

Comment:
I think this result should probably also exist as a `lemma StarOrderedRing.IsSelfAdjoint_of_nonneg`, in `Mathlib/Algebra/Star/Order.lean`

Reply:
I added it, but left it in `Positive.lean` because otherwise I would need to import `SelfAdjoint.lean` into `Order.lean`